src/runmode-dpdk.c

```diff
@@ -0,0 +1,1295 @@
+    .threads = "threads",
+    .promisc = "promisc",
+    .multicast = "multicast",
+    .checksum_checks = "checksum-checks",
+    .checksum_checks_offload = "checksum-checks-offload",
+    .mtu = "mtu",
+    .mempool_size = "mempool-size",
+    .mempool_cache_size = "mempool-cache-size",
+    .rx_descriptors = "rx-descriptors",
+    .tx_descriptors = "tx-descriptors",
+    .copy_mode = "copy-mode",
+    .copy_iface = "copy-iface",
+};
+
+char *AllocArgument(size_t arg_len)
+{
+    char *ptr;
+
+    arg_len += 1; // null character
+    ptr = (char *)SCCalloc(arg_len, sizeof(char));
+    if (ptr == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Could not allocate memory for an argument");
+        errno = ENOMEM;
+        SCReturnPtr(NULL, "char *");
+    }
+
+    SCReturnPtr(ptr, "char *");
+}
+
+/**
+ * Allocates space for length of the given string and then copies contents
+ * @param arg String to set to the newly allocated space
+ * @return memory address if no error otherwise NULL (with errno set)
+ */
+char *AllocAndSetArgument(const char *arg)
+{
+    if (arg == NULL) {
+        SCLogError(SC_ERR_DPDK_CONF, "Passed argument is NULL in DPDK config initialization");
+        errno = EINVAL;
+        SCReturnPtr(NULL, "char *");
+    }
+
+    char *ptr;
+    size_t arg_len = strlen(arg);
+
+    ptr = AllocArgument(arg_len);
+    if (ptr == NULL)
+        SCReturnPtr(NULL, "char *");
+
+    strlcpy(ptr, arg, arg_len + 1);
+    SCReturnPtr(ptr, "char *");
+}
+
+char *AllocAndSetOption(const char *arg)
+{
+    if (arg == NULL) {
+        errno = EINVAL;
+        SCReturnPtr(NULL, "char *");
+    }
+
+    char *ptr = NULL;
+    size_t arg_len = strlen(arg);
+    uint8_t is_long_arg = arg_len > 1;
+    const char *dash_prefix = is_long_arg ? "--" : "-";
+    size_t full_len = arg_len + strlen(dash_prefix);
+
+    ptr = AllocArgument(full_len);
+    if (ptr == NULL)
+        SCReturnPtr(NULL, "char *");
+
+    strlcpy(ptr, dash_prefix, strlen(dash_prefix) + 1);
+    strlcat(ptr, arg, full_len + 1);
+    SCReturnPtr(ptr, "char *");
+}
+
+int ArgumentsInit(struct Arguments *args, unsigned capacity)
+{
+    args->argv = SCCalloc(capacity, sizeof(args->argv));
+    if (args->argv == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Could not allocate memory for Arguments structure");
+        SCReturnInt(-ENOMEM);
+    }
+
+    args->capacity = capacity;
+    args->argc = 0;
+    SCReturnInt(0);
+}
+
+void ArgumentsCleanup(struct Arguments *args)
+{
+    for (int i = 0; i < args->argc; i++) {
+        if (args->argv[i] != NULL) {
+            SCFree(args->argv[i]);
+            args->argv[i] = NULL;
+        }
+    }
+
+    SCFree(args->argv);
+    args->argv = NULL;
+    args->argc = 0;
+    args->capacity = 0;
+}
+
+int ArgumentsAdd(struct Arguments *args, char *value)
+{
+    if (args->argc + 1 > args->capacity) {
+        SCLogError(SC_ERR_DPDK_EAL_INIT, "No capacity for more arguments");
+        SCReturnInt(-ENOBUFS);
```

Comment:
Based on this comment I have tried to rework the error handling. Currently I only set `errno` in functions where pointer is returned, in functions which return `int` I return negative value of the error code (e.g. `return -ENOMEM;`

Caller then checks if the result is negative and if so, it can propagate the value upwards or dump it.

Reply:
I believe this comment can be linked with [this comment](https://github.com/OISF/suricata/pull/6317#discussion_r715543800). If we agree on the use of FatalErrors I guess a lot of error handling goes away and thus it will be more consistent.